File: my_cpp_pkg/src/battery.cpp
#include "rclcpp/rclcpp.hpp"

#include "my_robot_interfaces/srv/set_led.hpp"

class BatteryNode : public rclcpp::Node // MODIFY NAME
{
public:
    BatteryNode() : Node("battery") // MODIFY NAME
    {
        timer_ = this->create_wall_timer(std::chrono::milliseconds(6000), std::bind(&BatteryNode::turnOnLed, this));
    }

    void turnOnLed()
    {
        // Create a thread to manage asyncronous request to the server
        thread1_ = std::thread(std::bind(&BatteryNode::callSetLedService,this,3,true));
        thread1_.detach();
        timer_->cancel();
        timer_ = this->create_wall_timer(std::chrono::milliseconds(4000), std::bind(&BatteryNode::turnOffLed, this));
    }

    void turnOffLed()
    {
        thread2_ = std::thread(std::bind(&BatteryNode::callSetLedService,this,3,false));
        thread2_.detach();
        timer_->cancel();
        timer_ = this->create_wall_timer(std::chrono::milliseconds(6000), std::bind(&BatteryNode::turnOnLed, this));
    }



    void callSetLedService(int led_number, bool state)
    {
        // Create the client
        auto client = this->create_client<my_robot_interfaces::srv::SetLed>("set_led");

        // Wait for the server to be up
        while(!client->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_WARN(this->get_logger(), "Waiting for the server to be up...");
        }

        // Create the request
        auto request = std::make_shared<my_robot_interfaces::srv::SetLed::Request>();
        request->led_number = led_number;
        request->state = state;

        // Send the request
        auto future = client->async_send_request(request);

        try{
            // Wait for the response
            auto response = future.get();

            RCLCPP_INFO(this->get_logger(), "Change led %d to state %s with a result of %s",
                        led_number,(state? "ON" : "OFF"),(response->success ? "Success" : "Failure"));
        }
        catch(const std::exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "Service call failed");
        }
    }



private:
    std::thread thread1_, thread2_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<BatteryNode>();   // MODIFY NAME
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: my_py_pkg/my_py_pkg/battery.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from functools import partial
from my_robot_interfaces.srv import SetLed

class BatteryNode(Node): 
    def __init__(self):
        super().__init__("battery") 

        # add timer
        self.timer_ = self.create_timer(4, self.set_led_on)

        

    def set_led_on(self):
        # Execute the client
        self.call_set_led_server(3,True)
        self.timer_.cancel()
        self.timer_ = self.create_timer(6, self.set_led_off)

    def set_led_off(self):
        # Execute the client
        self.call_set_led_server(3,False)
        self.timer_.cancel()
        self.timer_ = self.create_timer(4, self.set_led_on)

    def call_set_led_server(self, led_number, state):
        client = self.create_client(SetLed, "set_led")
    
        # Wait for the server
        while not client.wait_for_service(1.0):
            self.get_logger().warn("Waiting for Server set_led...")

        # Create a request object
        request = SetLed.Request()
        request.led_number = led_number
        request.state = state

        # Use async call as recommended
        future = client.call_async(request)

        # Add done callback to execute when future is complete
        # Use partial to add more arguments
        future.add_done_callback(partial(self.callback_call_set_led, led_number=led_number, state=state))

    def callback_call_set_led(self, future, led_number, state):
        try: 
            response = future.result()
            self.get_logger().info(f"Led number {str(led_number)} requested to {'ON' if state else 'OFF'} resulted in {'Success' if response.success else 'Failure'}")
        except Exception as e:
            self.get_logger().error("Service call failed %r" % (e,))



def main(args=None):
    rclpy.init(args=args)
    node = BatteryNode() 
    rclpy.spin(node)
    rclpy.shutdown()


if __name__=="__main__":
    main()